modules/imgcodecs/src/grfmt_jpeg2000_openjpeg.cpp

```diff
@@ -323,30 +323,28 @@ opj_dparameters setupDecoderParameters()
     return parameters;
 }
 
-opj_cparameters setupEncoderParameters(const std::vector<int>& params)
+opj_cparameters setupEncoderParameters(const std::vector<int>& params, const int numChannels)
 {
     opj_cparameters parameters;
     opj_set_default_encoder_parameters(&parameters);
-    bool rate_is_specified = false;
+    parameters.tcp_mct = numChannels >= 3 ? 1 : 0;
     for (size_t i = 0; i < params.size(); i += 2)
     {
         switch (params[i])
         {
         case cv::IMWRITE_JPEG2000_COMPRESSION_X1000:
-            parameters.tcp_rates[0] = 1000.f / std::min(std::max(params[i + 1], 1), 1000);
-            rate_is_specified = true;
+            if (params[i + 1] > 0)
+            {
+                parameters.tcp_rates[0] = 1000.f / std::min(params[i + 1], 1000);
+                parameters.tcp_numlayers = 1;
+                parameters.cp_disto_alloc = 1;
+            }
             break;
         default:
             CV_LOG_WARNING(NULL, "OpenJPEG2000(encoder): skip unsupported parameter: " << params[i]);
             break;
         }
     }
-    parameters.tcp_numlayers = 1;
-    parameters.cp_disto_alloc = 1;
-    if (!rate_is_specified)
-    {
-        parameters.tcp_rates[0] = 4;
-    }
```

Comment:
These default values helps to achieve reasonable quality in terms of PSNR metric with the smaller file size.
Now this idea has been thrown out of consideration.

As per original issue: only the change `parameters.tcp_mct = numChannels >= 3 ? 1 : 0;`  with provided `IMWRITE_JPEG2000_COMPRESSION_X1000` set to default value from `opj_compress` allow to achieve the same result, isn't it?

Reply:
Yes, you are right. Only the extenstion `parameters.tcp_mct = numChannels >= 3 ? 1 : 0;` and setting the rate to 1000 results in the same file as opj_compress. However, setting the rate `parameters.tcp_rates[0] = 4;` as long as it is not explicitly set differently results in a lossy compression by default. In particular, the resulting lossy image is the same as using opj_compress with `-r 4` option (which is therefore lossy as well), but with the important difference that the user explicitly enables the lossy compression while OpenCV does so by itself.

Here, my intention was only to obtain a consistent result with both, OpenCV and opj_compress. Maybe OpenCV could use different parameters by default, in particular if the file sizes are reduced. But if the codec offers lossless compression, using lossy by default in OpenCV is hardly satisfactory. Especially if the used library (i.e. OpenJPEG in this particular example) uses lossless compression by default.